File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceBindingCallback.java
/*
 * Copyright (c) 2021 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package org.forgerock.android.auth.callback;

import static androidx.biometric.BiometricManager.Authenticators.BIOMETRIC_STRONG;

import android.content.Context;
import android.os.Build;
import android.os.OperationCanceledException;

import androidx.annotation.Keep;
import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.biometric.BiometricPrompt;
import androidx.fragment.app.FragmentActivity;

import com.nimbusds.jose.JOSEException;
import com.nimbusds.jose.JWSAlgorithm;
import com.nimbusds.jose.JWSHeader;
import com.nimbusds.jose.crypto.RSASSASigner;
import com.nimbusds.jose.jwk.JWK;
import com.nimbusds.jose.jwk.KeyUse;
import com.nimbusds.jose.jwk.RSAKey;
import com.nimbusds.jwt.JWTClaimsSet;
import com.nimbusds.jwt.SignedJWT;

import org.forgerock.android.auth.DeviceIdentifier;
import org.forgerock.android.auth.FRListener;
import org.forgerock.android.auth.InitProvider;
import org.forgerock.android.auth.Listener;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.security.GeneralSecurityException;
import java.security.KeyStore;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.interfaces.RSAPublicKey;

import lombok.Getter;
import lombok.NoArgsConstructor;

@NoArgsConstructor
@Getter
public class DeviceBindingCallback extends AbstractCallback implements KeyAware {

    private String userId;
    private String challenge;
    private String authenticationType;
    private String title;
    private String subtitle;
    private String description;

    @Keep
    public DeviceBindingCallback(JSONObject jsonObject, int index) {
        super(jsonObject, index);
    }

    @Override
    protected void setAttribute(String name, Object value) {
        switch (name) {
            case "userId":
                this.userId = (String) value;
                break;
            case "challenge":
                this.challenge = (String) value;
                break;
            case "authenticationType":
                this.authenticationType = (String) value;
                break;
            case "title":
                this.title = (String) value;
                break;
            case "subtitle":
                this.subtitle = (String) value;
                break;
            case "description":
                this.description = (String) value;
                break;
            default:
                //ignore
        }
    }

    public void setJws(String value) {
        super.setValue(value, 0);
    }

    public void setDeviceName(String value) {
        super.setValue(value, 1);
    }

    public void setDeviceId(String value) {
        super.setValue(value, 2);
    }

    public void setClientError(String value) {
        super.setValue(value, 3);
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    public void bind(Context context, FRListener<Void> listener) {

        //Check if the device support it.
        if (!isSupported()) {
            setClientError("unsupported");
            Listener.onException(listener, new UnsupportedOperationException());
        }

        FragmentActivity fragmentActivity = InitProvider.getCurrentActivityAsFragmentActivity();

        String keyAlias;

        try {
            keyAlias = generateKeys(context);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }


        BiometricPrompt biometricPrompt = new BiometricPrompt(fragmentActivity, new BiometricPrompt.AuthenticationCallback() {
            @Override
            public void onAuthenticationError(int errorCode, @NonNull CharSequence errString) {
                super.onAuthenticationError(errorCode, errString);
                setClientError("unsupported");
                Listener.onException(listener, new OperationCanceledException(errString.toString()));
            }

            @Override
            public void onAuthenticationSucceeded(@NonNull BiometricPrompt.AuthenticationResult result) {
                super.onAuthenticationSucceeded(result);

                try {
                    String kid = getBindingInfo(context).getString("kid");
                    String jws = sign(keyAlias, challenge, kid);
                    setJws(jws);
                    setDeviceName("Andy Device");
                    setDeviceId(DeviceIdentifier.builder().context(context).build().getIdentifier());
                    Listener.onSuccess(listener, null);
                } catch (Exception e) {
                    Listener.onException(listener, e);
                }
            }

            @Override
            public void onAuthenticationFailed() {
                super.onAuthenticationFailed();
                setClientError("unsupported");
                Listener.onException(listener, new OperationCanceledException());
            }
        });

        BiometricPrompt.PromptInfo promptInfo = new BiometricPrompt.PromptInfo.Builder()
                .setTitle(getTitle())
                .setSubtitle(getSubtitle())
                .setConfirmationRequired(true)
                .setDescription(getDescription())
                .setNegativeButtonText("Cancel")
                .setConfirmationRequired(true)
                //.setAllowedAuthenticators(BIOMETRIC_STRONG | DEVICE_CREDENTIAL).build();
                .setAllowedAuthenticators(BIOMETRIC_STRONG).build();

        biometricPrompt.authenticate(promptInfo);

    }

    protected boolean isSupported() {
        return true;
    }

    @Override
    public String getType() {
        return DeviceBindingCallback.class.getSimpleName();
    }


    @RequiresApi(api = Build.VERSION_CODES.M)
    private String sign(String keyAlias, String challenge, String kid) throws GeneralSecurityException, IOException, JSONException, JOSEException {

        KeyStore keyStore = getKeyStore();
        PrivateKey privateKey = (PrivateKey) keyStore.getKey(keyAlias, null);

        JWK jwk = new RSAKey.Builder((RSAPublicKey) getPublicKey(keyAlias))
                .keyUse(KeyUse.SIGNATURE)
                .keyID(kid)
                .algorithm(JWSAlgorithm.RS512)
                .build();

        SignedJWT signedJWT = new SignedJWT(
                new JWSHeader.Builder(JWSAlgorithm.RS512)
                        .keyID(kid)
                        .jwk(jwk)
                        .build()
                , new JWTClaimsSet.Builder()
                .subject(getUserId())
                .claim("challenge", challenge)
                .build());

        signedJWT.sign(new RSASSASigner(privateKey));
        return signedJWT.serialize();

    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    private PublicKey getPublicKey(String keyAlias) throws GeneralSecurityException, IOException {
        KeyStore keyStore = getKeyStore();
        return keyStore.getCertificate(keyAlias).getPublicKey();
    }

}

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceBindingCallback.kt
/*
 * Copyright (c) 2022 - 2024 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
package org.forgerock.android.auth.callback

import android.content.Context
import android.util.Base64
import androidx.annotation.Keep
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeout
import org.forgerock.android.auth.CryptoKey
import org.forgerock.android.auth.DeviceIdentifier
import org.forgerock.android.auth.FRListener
import org.forgerock.android.auth.Listener
import org.forgerock.android.auth.devicebind.ApplicationPinDeviceAuthenticator
import org.forgerock.android.auth.devicebind.BiometricAndDeviceCredential
import org.forgerock.android.auth.devicebind.BiometricAuthenticator
import org.forgerock.android.auth.devicebind.BiometricOnly
import org.forgerock.android.auth.devicebind.DeviceAuthenticator
import org.forgerock.android.auth.devicebind.DeviceBindingErrorStatus
import org.forgerock.android.auth.devicebind.DeviceBindingErrorStatus.*
import org.forgerock.android.auth.devicebind.DeviceBindingException
import org.forgerock.android.auth.devicebind.DeviceBindingRepository
import org.forgerock.android.auth.devicebind.DeviceBindingStatus
import org.forgerock.android.auth.devicebind.KeyPair
import org.forgerock.android.auth.devicebind.LocalDeviceBindingRepository
import org.forgerock.android.auth.devicebind.None
import org.forgerock.android.auth.devicebind.Prompt
import org.forgerock.android.auth.devicebind.Success
import org.forgerock.android.auth.devicebind.UserKey
import org.forgerock.android.auth.devicebind.initialize
import org.json.JSONObject
import java.util.*

/**
 * Callback to collect the device binding information
 */
open class DeviceBindingCallback : AbstractCallback, Binding {

    @Keep
    @JvmOverloads
    constructor(jsonObject: JSONObject, index: Int) : super(jsonObject, index)

    @Keep
    @JvmOverloads
    constructor() : super()

    /**
     * The userId received from server
     */
    lateinit var userId: String
        private set

    /**
     * The userName received from server
     */
    lateinit var userName: String
        private set

    /**
     * The challenge received from server
     */
    lateinit var challenge: String
        private set

    /**
     * The authentication type of the journey
     */
    lateinit var deviceBindingAuthenticationType: DeviceBindingAuthenticationType
        private set

    /**
     * The title to be displayed in biometric prompt
     */
    lateinit var title: String
        private set

    /**
     * The subtitle to be displayed in biometric prompt
     */
    lateinit var subtitle: String
        private set

    /**
     * The description to be displayed in biometric prompt
     */
    lateinit var description: String
        private set

    /**
     * The timeout to be to expire the biometric authentication
     */
    var timeout: Int? = null
        private set

    /**
     * Enable Attestation
     */
    lateinit var attestation: Attestation
        private set

    init {
        //If attestation is not provided, default to NONE
        if (!::attestation.isInitialized) {
            attestation = Attestation.None
        }
    }

    final override fun setAttribute(name: String, value: Any) = when (name) {
        "userId" -> userId = value as String
        "username" -> userName = value as String
        "challenge" -> challenge = value as String
        "authenticationType" -> deviceBindingAuthenticationType =
            DeviceBindingAuthenticationType.valueOf(value as String)
        "title" -> title = value as? String ?: ""
        "subtitle" -> subtitle = value as? String ?: ""
        "description" -> description = value as? String ?: ""
        "timeout" -> timeout = value as? Int
        "attestation" -> attestation = Attestation
            .fromBoolean(value as Boolean, Base64.decode(challenge, Base64.NO_WRAP))
        else -> {}
    }

    override fun getType(): String {
        return "DeviceBindingCallback"
    }

    /**
     * Input the JWS key to the server
     * @param value The JWS value.
     */
    fun setJws(value: String?) {
        super.setValue(value, 0)
    }

    /**
     * Input the Device Name to the server
     * @param value The device name value.
     */
    fun setDeviceName(value: String?) {
        super.setValue(value, 1)
    }

    /**
     * Input the Device Id to the server
     * @param value The device Id.
     */
    fun setDeviceId(value: String?) {
        super.setValue(value, 2)
    }

    /**
     * Input the Client Error to the server
     * @param value DeviceBind ErrorType .
     */
    override fun setClientError(value: String?) {
        super.setValue(value, 3)
    }

    /**
     * Bind the device. Calling the [bind] function, the existing bounded keys will be removed.
     * If don't want to replace or remove existing keys, please use [FRUserKeys] to check existing
     * keys before calling this method
     *
     * @param context  The Application Context
     * @param deviceAuthenticator A function to return a [DeviceAuthenticator], [deviceAuthenticatorIdentifier] will be used if not provided
     * @param listener The Listener to listen for the result
     * @param prompt The Prompt to modify the title, subtitle, description
     */
    @JvmOverloads
    open fun bind(context: Context,
                  deviceAuthenticator: (type: DeviceBindingAuthenticationType) -> DeviceAuthenticator = deviceAuthenticatorIdentifier,
                  listener: FRListener<Void?>,
                  prompt: Prompt? = null) {
        val scope = CoroutineScope(Dispatchers.Default)
        scope.launch {
            try {
                bind(context,  prompt = prompt, deviceAuthenticator = deviceAuthenticator)
                Listener.onSuccess(listener, null)
            } catch (e: Exception) {
                Listener.onException(listener, e)
            }
        }
    }

    /**
     * Bind the device. Calling the [bind] function, the existing bounded keys will be removed.
     * If don't want to replace or remove existing keys, please use [FRUserKeys] to check existing
     * keys before calling this method
     *
     * @param context  The Application Context
     * @param prompt The Prompt to modify the title, subtitle, description
     * @param deviceAuthenticator A function to return a [DeviceAuthenticator], [deviceAuthenticatorIdentifier] will be used if not provided
     */
    open suspend fun bind(context: Context,
                          prompt: Prompt? = null,
                          deviceAuthenticator: (type: DeviceBindingAuthenticationType) -> DeviceAuthenticator = deviceAuthenticatorIdentifier) {
        execute(context, deviceAuthenticator(deviceBindingAuthenticationType), prompt = prompt)
    }


    /**
     * Helper method to execute binding , signing, show biometric prompt.
     *
     * @param context  The Application Context
     * @param deviceAuthenticator Interface to find the Authentication Type
     * @param deviceBindingRepository Persist the values in encrypted shared preference
     * @param deviceId Generated Device Identifier
     * @param prompt The Prompt to modify the title, subtitle, description
     */
    internal suspend fun execute(context: Context,
                                 deviceAuthenticator: DeviceAuthenticator = getDeviceAuthenticator(
                                     deviceBindingAuthenticationType),
                                 deviceBindingRepository: DeviceBindingRepository = LocalDeviceBindingRepository(
                                     context),
                                 deviceId: String = DeviceIdentifier.builder().context(context)
                                     .build().identifier,
                                 prompt: Prompt? = null) {

        deviceAuthenticator.initialize(userId, prompt ?: Prompt(title, subtitle, description))

        if (deviceAuthenticator.isSupported(context, attestation).not()) {
            handleException(DeviceBindingException(Unsupported()))
            return
        }

        var keyPair: KeyPair?
        var userKey: UserKey? = null
        try {
            val status: DeviceBindingStatus
            withTimeout(getDuration(timeout)) {
                clearKeys(context, deviceAuthenticator)
                keyPair =
                    deviceAuthenticator.generateKeys(context, attestation)
                status = deviceAuthenticator.authenticate(context)
            }
            when (status) {
                is Success -> {
                    keyPair?.let { kp ->
                        userKey = UserKey(kp.keyAlias,
                            userId, userName, kid = UUID.randomUUID().toString(),
                            deviceBindingAuthenticationType
                        )
                        userKey?.let {
                            deviceBindingRepository.persist(it)
                            val jws = deviceAuthenticator.sign(context,
                                kp,
                                status.signature,
                                it.kid,
                                userId,
                                challenge,
                                getExpiration(timeout),
                                attestation)
                            setJws(jws)
                            setDeviceId(deviceId)
                        }
                    }
                }
                is DeviceBindingErrorStatus -> {
                    throw DeviceBindingException(status)
                }
            }
        } catch (e: Exception) {
            userKey?.let { deviceBindingRepository.delete(it) }
            deviceAuthenticator.deleteKeys(context)
            handleException(e)
        }
    }

    /**
     * For now we don't support multiple keys, so before we create new keys,
     * we clean existing keys.
     */
    private fun clearKeys(context: Context, deviceAuthenticator: DeviceAuthenticator) {
        when (deviceAuthenticator) {
            is ApplicationPinDeviceAuthenticator -> {
                //Delete Keys from keystore
                getCryptoKey().deleteKeys()
            }
            is BiometricAuthenticator, is None -> {
                ApplicationPinDeviceAuthenticator().initialize(userId).deleteKeys(context)
            }
        }
        deviceAuthenticator.deleteKeys(context)
    }

    open fun getCryptoKey() = CryptoKey(userId)
}


/**
 * convert authentication string received from server to authentication enum
 */
enum class DeviceBindingAuthenticationType constructor(val serializedValue: String?) {
    BIOMETRIC_ONLY("BIOMETRIC_ONLY"), BIOMETRIC_ALLOW_FALLBACK("BIOMETRIC_ALLOW_FALLBACK"), NONE("NONE"), APPLICATION_PIN(
        "APPLICATION_PIN");
}

fun DeviceBindingAuthenticationType.getAuthType(): DeviceAuthenticator {
    return when (this) {
        DeviceBindingAuthenticationType.BIOMETRIC_ONLY -> BiometricOnly()
        DeviceBindingAuthenticationType.APPLICATION_PIN -> ApplicationPinDeviceAuthenticator()
        DeviceBindingAuthenticationType.BIOMETRIC_ALLOW_FALLBACK -> BiometricAndDeviceCredential()
        else -> None()
    }
}

